src/Bicep.Core/Emit/SourceMapHelper.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.Workspaces;
+using Microsoft.WindowsAzure.ResourceStack.Common.Extensions;
+
+namespace Bicep.Core.Emit
+{
+    public static class SourceMapHelper
+    {
+        public static void AddMapping(
+            IDictionary<string, IDictionary<int, IList<(int start, int end)>>> rawSourceMap,
+            BicepFile bicepFile,
+            SyntaxBase bicepSyntax,
+            PositionTrackingJsonTextWriter jsonWriter,
+            int jsonStartPos)
+        {
+            if (rawSourceMap is null)
+            {
+                throw new ArgumentNullException(nameof(rawSourceMap));
+            }
+
+            var bicepFilePath = bicepFile.FileUri.AbsolutePath;
+            (int bicepLine, _) = TextCoordinateConverter.GetPosition(bicepFile.LineStarts, bicepSyntax.GetPosition());
+
+            // account for leading nodes (decorators)
+            if (bicepSyntax is StatementSyntax syntax)
+            {
+                bicepLine += syntax.LeadingNodes.Count(node => node is Token token && token.Type is TokenType.NewLine);
```

Comment:
This isn't going to be robust - the newline token can span multiple newlines, and certain pieces of syntax (multiline comments, verbatim strings) can span multiple lines. You should be able to instead do a textspan calculation to get an accurate position.

You'll want something like the following (pseudocode):
```
var last = leadingnodes.lastordefault();
if (last is null)
  // no leading nodes - use the full span
  return syntax.Span;
else
  // get span between end of last leading node and end of statement
  return new TextSpan(last.position + last.length, syntax.length - last.length);
```